### PR TITLE
implemented applying markup bonuses to parent agency

### PR DESCRIPTION
**Pull request by DFSko**
happy-travel/agent-app-project#9

### CODE REVIEW

File: Api/Services/Markups/MarkupBonusMaterializationService.cs [hunk 1]
```diff
@@ -64,62 +58,111 @@ private Task<List<MaterializationData>> GetData(ICollection<int> markupsForMater
...
+                    select parentAgency.Id;
+
+                var parentAgencyId = await parentAgencyQuery.SingleOrDefaultAsync();
+                await Apply(data.PolicyId, data.ReferenceCode, parentAgencyId, data.Amount);
+            }
         }
 
 
-        private async Task WriteLog(MaterializationData data)
+        private async Task Apply(int policyId, string referenceCode, int agencyId, decimal amount)
```
Comment (mingazhev) [inline Api/Services/Markups/MarkupBonusMaterializationService.cs:106]: Why just Apply, not ApplyAgencyBonus?

File: Api/Services/Markups/MarkupBonusMaterializationService.cs [hunk 2]
```diff
@@ -64,62 +58,111 @@ private Task<List<MaterializationData>> GetData(ICollection<int> markupsForMater
...
-            _context.MaterializationBonusLogs.Add(new MaterializationBonusLog
+            var agencyAccount = await _context.AgencyAccounts
+                .SingleOrDefaultAsync(a => a.AgencyId == agencyId);
+                
+            if(agencyAccount is null)
+                return;
+
+            var paidDate = _dateTimeProvider.UtcNow();
+            
+            await using (var transaction = await _context.Database.BeginTransactionAsync())
```
Comment (mingazhev) [inline Api/Services/Markups/MarkupBonusMaterializationService.cs:116]: We have extension methods that uses transactions with using execution strategy. I think you should use them if possible

File: Api/Services/Markups/MarkupBonusMaterializationService.cs [hunk 3]
```diff
@@ -64,62 +58,111 @@ private Task<List<MaterializationData>> GetData(ICollection<int> markupsForMater
...
 
 
-        private async Task WriteLog(MaterializationData data)
+        private async Task Apply(int policyId, string referenceCode, int agencyId, decimal amount)
         {
-            _context.MaterializationBonusLogs.Add(new MaterializationBonusLog
+            var agencyAccount = await _context.AgencyAccounts
+                .SingleOrDefaultAsync(a => a.AgencyId == agencyId);
+                
+            if(agencyAccount is null)
```
Comment (mingazhev) [inline Api/Services/Markups/MarkupBonusMaterializationService.cs:111]: Whitespace missing?

**Review by mingazhev — COMMENTED** [review]
And please add unit tests for the whole class. This functionality is important but very hardly testable.

File: Api/Services/Markups/MarkupBonusMaterializationService.cs (hunk 2 above)
Comment (Raibo) [reply Api/Services/Markups/MarkupBonusMaterializationService.cs:116]: BindWithTransaction

**Review by kirillta — APPROVED** [review]

File: Api/Services/Markups/MarkupBonusMaterializationService.cs [hunk 4]
```diff
@@ -64,62 +58,111 @@ private Task<List<MaterializationData>> GetData(ICollection<int> markupsForMater
...
-        {
-            var appliedMarkup = await _context.AppliedBookingMarkups
-                .SingleOrDefaultAsync(a => a.PolicyId == data.PolicyId && a.ReferenceCode == data.ReferenceCode);
-
-            appliedMarkup.Paid = _dateTimeProvider.UtcNow();
-            _context.AppliedBookingMarkups.Update(appliedMarkup);
-            await _context.SaveChangesAsync();
-            _context.Entry(appliedMarkup).State = EntityState.Detached;
-            return data;
+            Task ApplyAgentBonus() => Apply(data.PolicyId, data.ReferenceCode, data.AgencyId, data.Amount);
```
[the remark this answers is not in the document]
Comment (DFSko) [reply Api/Services/Markups/MarkupBonusMaterializationService.cs]: Fixed

File: Api/Services/Markups/MarkupBonusMaterializationService.cs (hunk 2 above)
Comment (DFSko) [reply Api/Services/Markups/MarkupBonusMaterializationService.cs]: Fixed

File: Api/Services/Markups/MarkupBonusMaterializationService.cs (hunk 3 above)
Comment (DFSko) [reply Api/Services/Markups/MarkupBonusMaterializationService.cs]: Fixed

File: Api/Services/Markups/MarkupBonusMaterializationService.cs (hunk 1 above)
Comment (DFSko) [reply Api/Services/Markups/MarkupBonusMaterializationService.cs]: Renamed

File: Api/Models/Markups/MaterializationData.cs [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
+using HappyTravel.Edo.Common.Enums.Markup;
 
 namespace HappyTravel.Edo.Api.Models.Markups
 {
     public readonly struct MaterializationData
     {
         public int PolicyId { get; init; }
         public string ReferenceCode { get; init; }
-        public int AgencyAccountId { get; init; }
+        public int AgencyId { get; init; }
```
[the remark this answers is not in the document]
Comment (DFSko) [reply Api/Models/Markups/MaterializationData.cs:11]: Agency accounts retrieved later

File: Api/Services/Markups/MarkupBonusMaterializationService.cs [hunk 6]
```diff
@@ -64,62 +71,114 @@ private Task<List<MaterializationData>> GetData(ICollection<int> markupsForMater
...
+            Task<Result> ApplyAgentScopeBonus() 
+                => ApplyAgencyBonus(data.PolicyId, data.ReferenceCode, data.AgencyId, data.Amount);
+
+
+            async Task<Result> ApplyAgencyScopeBonus()
+            {
+                var parentAgencyQuery = from agency in _context.Agencies
+                    join parentAgency in _context.Agencies on agency.ParentId equals parentAgency.Id
+                    where agency.Id == data.AgencyId
+                    select parentAgency.Id;
```
Comment (mingazhev) [inline Api/Services/Markups/MarkupBonusMaterializationService.cs:116]: Could this be done easier? Just getting an agency by id?

Comment (DFSko) [reply Api/Services/Markups/MarkupBonusMaterializationService.cs]: Simplified

**Review by mingazhev — APPROVED** [review]

**Pull request merged**